### ISSUE TITLE
Reflexive has_one through with proc issue

### ISSUE THREAD

**Issue by jith51** [rails/rails#22169]
Hi,
I recently posted on the issue [6663](https://github.com/rails/rails/issues/6663). My post extended the issue concerning has_one through statement with STI to has_one through statement with proc and reflexive class.

As this isssue is closed, I open a new issue and join the report_bug_template below.

``` ruby
begin
  require 'bundler/inline'
rescue LoadError => e
  $stderr.puts 'Bundler version 1.10 or later is required. Please update your Bundler'
  raise e
end

gemfile(true) do
  source 'https://rubygems.org'
  # Activate the gem you are reporting the issue against.
  gem 'activerecord', '4.2.3'
  gem 'sqlite3'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :people, force: true  do |t|
    t.string :gender
    t.references :child
  end
end

class Person < ActiveRecord::Base
  has_one :mother, ->{where(gender: 'F')}, class_name: 'Person', foreign_key: :child_id
  has_one :father, ->{where(gender: 'M')}, class_name: 'Person', foreign_key: :child_id
  has_one :father_s_mother, ->{where(gender: 'F')}, through: :father, source: :mother
end


class BugTest < Minitest::Test
  def test_association_chip_in_computer
    mother = Person.create(gender:'F')
    father = Person.create(gender:'M',mother: mother)

    child=Person.create(father:father)

    assert_equal mother.id, father.mother.try(:id)
    assert_equal father.id, child.father.try(:id)
    assert_equal mother.id, child.father_s_mother.try(:id)
  end
end
```

Below the wrong generated SQL for the last assertion : 

``` SQL
    SELECT "people".* 
      FROM "people"
INNER JOIN "people" "fathers_father_s_mother_join"
        ON "people"."child_id" = "fathers_father_s_mother_join"."id"
     WHERE "fathers_father_s_mother_join"."child_id" = ?
       AND "people"."gender" = ?
```

``` SQL
       AND "people"."gender" = ?
       AND "people"."gender" = ?
```

``` SQL
    LIMIT 1 [["parent_id", 6], ["gender", "F"], ["gender", "F"], ["gender", "M"]]
```

And the expected SQL : 

``` SQL
    SELECT "people".* 
      FROM "people"
INNER JOIN "people" "fathers_father_s_mother_join"
        ON "people"."child_id" = "fathers_father_s_mother_join"."id"
     WHERE "fathers_father_s_mother_join"."child_id" = ?
       AND "people"."gender" = ?
```

``` SQL
       AND "fathers_father_s_mother_join"."gender" = ?
```

``` SQL
    LIMIT 1 [["parent_id", 6],  ["gender", "F"], ["gender", "M"]]
```

**Comment by pixeltrix** [comment]
@jith51 is this the same as #22157 that you opened yesterday?

**Comment by jith51** [comment]
Hi,

Both concern has_one through relation and bad aliasing but : 
- The first https://github.com/rails/rails/issues/22157 concern STI association
- This issue concern reflexive association with PROC (scope)

I have looked in the code and perhaps the both issues come from the main source. I think the method target_scope line 14 in [through_association.rb](https://github.com/rails/rails/blob/7f18ea14c893cb5c9f04d4fda9661126758332b5/activerecord/lib/active_record/associations/through_association.rb) is probably the guilty. It seems not to use the alias_tracker constructing the SQL clause for default_scope conditions and STI conditions. 

But I am not sure. So I created two issues.
